templates/cisco_nxos_show_forwarding_adjacency.textfsm

```diff
@@ -0,0 +1,30 @@
+# Nexus 7k have multiple slots
+Value Filldown SLOT (\d+)
+Value Required NEXTHOP ((?:\d{1,3}\.){3}\d{1,3})
+Value Required REWRITE (\S+)
+Value Required INTERFACE (\S+)
+Value ORIGIN_AS (\S+)
+Value PEER_AS (\S+)
+Value NEIGHBOR (\S+)
+
+Start
+  # Match headers
+  ^IPv4\s+adjacency\s+information
+  ^next-hop\s+rewrite\s+info\s+interface
+  # Extract slot, if available (Nexus 7k)
+  ^slot\s+${SLOT}
+  ^====
+  # List of entries detected
+  ^-------- -> Entries
```

Comment:
Yep. Let me review, need to find some output from multi-chassis N7K

Reply:
I've analyzed few more outputs from N7K and came up with the following:

```
# Nexus 7k have multiple slots
Value Filldown SLOT (\d+)
Value Required NEXTHOP ((?:\d{1,3}\.){3}\d{1,3})
Value Required REWRITE (\S+)
Value Required INTERFACE (\S+)
Value ORIGIN_AS (\S+)
Value PEER_AS (\S+)
Value NEIGHBOR (\S+)

Start
  # Extract slot #, if available (Nexus 7k)
  ^slot\s+${SLOT}
  ^====
  # Match headers
  ^IPv4\s+adjacency\s+information
  ^next-hop\s+rewrite\s+info\s+interface\s*$$ -> Entries
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

Entries
  # Ignore delimiter
  ^--------
  # Match single adjacency entry
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error
```

Even though I do not split different entries (brief vs detail) into different states, I do capture slot and re-run parser from scratch if slot is detected. Also, I removed one parsing line as origin_as, peer_as and neighbor are three optional attributes and they can be captured as per above.

Alternative would be the following (as advised by yourself), but IMHO it's an overkill... as there's a lot of repitative code :) I will however leave it to your final say.

```
# Nexus 7k have multiple slots
Value Filldown SLOT (\d+)
Value Required NEXTHOP ((?:\d{1,3}\.){3}\d{1,3})
Value Required REWRITE (\S+)
Value Required INTERFACE (\S+)
Value ORIGIN_AS (\S+)
Value PEER_AS (\S+)
Value NEIGHBOR (\S+)

Start
  # Extract slot #, if available (Nexus 7k)
  ^slot\s+${SLOT}
  ^====
  # Match headers
  ^IPv4\s+adjacency\s+information
  ^next-hop\s+rewrite\s+info\s+interface\s*$$ -> Entries
  ^next-hop\s+rewrite\s+info\s+interface\s+Origin\s+AS\s+Peer\s+AS\s+Neighbor -> EntriesDetail
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

Entries
  # Ignore delimiter
  ^--------
  # Match single adjacency entry (brief view)
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE} -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

EntriesDetail
  # Ignore delimiter
  ^--------
  # Match single adjacency entry (detailed view)
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error
```

As you can see, this line
`^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record` is required in both variants because even if it's detailed view those three attributes are still optional and we need to account for that.

Reply:
I've analyzed few more outputs from N7K and came up with the following:

```
# Nexus 7k have multiple slots
Value Filldown SLOT (\d+)
Value Required NEXTHOP ((?:\d{1,3}\.){3}\d{1,3})
Value Required REWRITE (\S+)
Value Required INTERFACE (\S+)
Value ORIGIN_AS (\S+)
Value PEER_AS (\S+)
Value NEIGHBOR (\S+)

Start
  # Extract slot #, if available (Nexus 7k)
  ^slot\s+${SLOT}
  ^====
  # Match headers
  ^IPv4\s+adjacency\s+information
  ^next-hop\s+rewrite\s+info\s+interface\s*$$ -> Entries
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

Entries
  # Ignore delimiter
  ^--------
  # Match single adjacency entry
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error
```

Even though I do not split different entries (brief vs detail) into different states, I do capture slot and re-run parser from scratch if slot is detected. Also, I removed one parsing line as origin_as, peer_as and neighbor are three optional attributes and they can be captured as per above.

Alternative would be the following (as advised by yourself), but IMHO it's an overkill... as there's a lot of repitative code :) I will however leave it to your final say.

```
# Nexus 7k have multiple slots
Value Filldown SLOT (\d+)
Value Required NEXTHOP ((?:\d{1,3}\.){3}\d{1,3})
Value Required REWRITE (\S+)
Value Required INTERFACE (\S+)
Value ORIGIN_AS (\S+)
Value PEER_AS (\S+)
Value NEIGHBOR (\S+)

Start
  # Extract slot #, if available (Nexus 7k)
  ^slot\s+${SLOT}
  ^====
  # Match headers
  ^IPv4\s+adjacency\s+information
  ^next-hop\s+rewrite\s+info\s+interface\s*$$ -> Entries
  ^next-hop\s+rewrite\s+info\s+interface\s+Origin\s+AS\s+Peer\s+AS\s+Neighbor -> EntriesDetail
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

Entries
  # Ignore delimiter
  ^--------
  # Match single adjacency entry (brief view)
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE} -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error

EntriesDetail
  # Ignore delimiter
  ^--------
  # Match single adjacency entry (detailed view)
  ^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record
  # New slot?
  ^slot\s+${SLOT} -> Start
  # Empty and unknowns lines
  ^\s*$$
  ^. -> Error
```

As you can see, this line
`^${NEXTHOP}\s+${REWRITE}\s+${INTERFACE}\s*(?:${ORIGIN_AS}\s+${PEER_AS}\s+${NEIGHBOR})? -> Record` is required in both variants because even if it's detailed view those three attributes are still optional and we need to account for that.